File: pb/src/main/java/org/sqlbuilder/pb/PbModule.java
package org.sqlbuilder.pb;

import org.sqlbuilder.common.Module;
import org.sqlbuilder.pb.collectors.PbExportCollector;
import org.sqlbuilder.pb.collectors.PbCollector;
import org.sqlbuilder.pb.collectors.PbUpdateCollector;

import java.io.IOException;

public class PbModule extends Module
{
	public static final String MODULE_ID = "pb";

	protected PbModule(final String conf, final Mode mode)
	{
		super(MODULE_ID, conf, mode);
	}

	@Override
	protected void run()
	{
		assert props != null;

		switch (mode)
		{
			case PLAIN:
			case RESOLVE:
				new PbCollector(props).run();
				try
				{
					Inserter inserter = mode == Mode.PLAIN ? new Inserter(props) : new ResolvingInserter(props);
					inserter.insert();
				}
				catch (IOException | ClassNotFoundException e)
				{
					e.printStackTrace();
				}
				break;

			case UPDATE:
				new PbUpdateCollector(props).run();
					try
				{
					Inserter inserter = new ResolvingUpdater(props);
					inserter.insert();
				}
				catch (IOException | ClassNotFoundException e)
				{
					e.printStackTrace();
				}
				break;

			case EXPORT:
				new PbExportCollector(props).run();
				try
				{
					Exporter exporter = new Exporter(props);
					exporter.run();
				}
				catch (IOException e)
				{
					e.printStackTrace();
				}
				break;

			default:
		}
	}

	public static void main(final String[] args)
	{
		int i = 0;
		Mode mode = Mode.PLAIN;
		if (args[i].startsWith("-"))
		{
			mode = Mode.read(args[i++]);
		}
		String conf = args[i];
		new PbModule(conf, mode).run();
	}
}

File: pb/src/main/java/org/semantikos/pb/PbModule.kt
package org.semantikos.pb

import org.semantikos.common.Module
import org.semantikos.pb.collectors.PbCollector
import org.semantikos.pb.collectors.PbExportCollector
import org.semantikos.pb.collectors.PbUpdateCollector
import java.io.IOException

open class PbModule protected constructor(conf: String, mode: Mode) : Module(MODULE_ID, conf, mode) {

    override fun run() {
        checkNotNull(props)

        when (mode) {
            Mode.PLAIN, Mode.RESOLVE -> {
                PbCollector(props).run()
                try {
                    val inserter = if (mode == Mode.PLAIN) Inserter(props) else ResolvingInserter(props)
                    inserter.insert()
                } catch (e: IOException) {
                    e.printStackTrace()
                } catch (e: ClassNotFoundException) {
                    e.printStackTrace()
                }
            }

            Mode.UPDATE              -> {
                PbUpdateCollector(props).run()
                try {
                    val inserter: Inserter = ResolvingUpdater(props)
                    inserter.insert()
                } catch (e: IOException) {
                    e.printStackTrace()
                } catch (e: ClassNotFoundException) {
                    e.printStackTrace()
                }
            }

            Mode.EXPORT              -> {
                PbExportCollector(props).run()
                try {
                    val exporter = Exporter(props)
                    exporter.run()
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }

            else                     -> {}
        }
    }

    companion object {

        const val MODULE_ID: String = "pb"

        @JvmStatic
        fun main(args: Array<String>) {
            var i = 0
            var mode: Mode = Mode.PLAIN
            if (args[i].startsWith("-")) {
                mode = Mode.read(args[i++])
            }
            val conf: String = args[i]
            PbModule(conf, mode).run()
        }
    }
}
